qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -150,30 +124,58 @@ def run(self, dag):
         Returns:
             DAGCircuit: the optimized DAG.
         """
-        if self._decomposers is None:
-            logger.info("Skipping pass because no basis is set")
+        if self._basis_gates is None and self._target is None:
+            logger.info("Skipping pass because no basis or target is set")
             return dag
 
         runs = dag.collect_1q_runs()
         for run in runs:
-            # SPECIAL CASE: Don't bother to optimize single U3 gates which are in the basis set.
-            #     The U3 decomposer is only going to emit a sequence of length 1 anyhow.
-            if "u3" in self._target_basis and len(run) == 1 and isinstance(run[0].op, U3Gate):
-                # Toss U3 gates equivalent to the identity; there we get off easy.
-                if np.allclose(run[0].op.to_matrix(), np.eye(2), 1e-15, 0):
-                    dag.remove_op_node(run[0])
-                    continue
-                # We might rewrite into lower `u`s if they're available.
-                if "u2" not in self._target_basis and "u1" not in self._target_basis:
-                    continue
-
-            new_basis, new_circ = self._resynthesize_run(run)
-
-            if new_circ is not None and self._substitution_checks(dag, run, new_circ, new_basis):
+            qubit = dag.qubits.index(run[0].qargs[0])
+            new_circ = self._resynthesize_run(run, qubit)
+
+            if self._target is None:
+                basis = self._basis_gates
+            else:
+                basis = self._target.operation_names_for_qargs((qubit,))
+
+            if new_circ is not None and self._substitution_checks(dag, run, new_circ, basis, qubit):
                 new_dag = circuit_to_dag(new_circ)
                 dag.substitute_node_with_dag(run[0], new_dag)
                 # Delete the other nodes in the run
                 for current_node in run[1:]:
                     dag.remove_op_node(current_node)
 
         return dag
+
+
+def _possible_decomposers(basis_set):
+    decomposers = []
+    euler_basis_gates = one_qubit_decompose.ONE_QUBIT_EULER_BASIS_GATES
+    for euler_basis_name, gates in euler_basis_gates.items():
+        if set(gates).issubset(basis_set):
+            decomposer = one_qubit_decompose.OneQubitEulerDecomposer(euler_basis_name)
+            decomposers.append(decomposer)
+    return decomposers
+
+
+def _error(circuit, target, qubit):
+    """
+    Calculate a rough error for a `circuit` that runs on a specific
+    `qubit` of `target` (circuit could also be a list of DAGNodes)
+
+    Use basis errors from target if available, otherwise use length
+    of circuit as a weak proxy for error.
+    """
+    if target is None:
+        return len(circuit)
+    else:
+        if isinstance(circuit, list):
+            gate_errors = [
+                1 - getattr(target[node.name].get((qubit,)), "error", 0.0) for node in circuit
```

Comment:
@mtreinish your fix in #8925 partly fixed the issue but I still had to work around something here. I'm not sure if this is still considered a bug or not.

For a global gate like `'rz'` even though it semantically means that it exists on qubit `(0,)`, still `target['rz'][(0,)]` raises a KeyError. Maybe bare access to the Target as a dict should not be allowed, don't know. But also there's no alternative method to get the properties of such global gate on some qubit, so I resorted to `dict.get()` here.

Reply:
I think the only way to avoid the `get()` usage is to add the api we were talking about in #8892. Since the internal dictionaries can have `None` as the payloads for the global ideal gate case (both for qargs and the properties) I'm not sure there is a way around that without a dedicated helper api that does the handling of `None` for us. I'll work on adding the api for #8892 now

Reply:
I added these function in: https://github.com/Qiskit/qiskit-terra/pull/9158